### PR TITLE
[REVIEW] concatenate row items using a separator defined per row

### LINKED ISSUE #3726: [FEA] Concatenate column to scalar and scalar to column
The string column concatenate function -- found in https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/include/cudf/strings/combine.hpp -- currently supports concatenation of multiple string columns. However, it does not support concatenating the same string to every row in the string column (appending or prepending). This is specifically relevant to replicating Spark's concat string functionality.

Add append and prepend functions.
```
std::unique_ptr<column> append(column_view const& string_column,
                               string_scalar const& value,
                               string_scalar const& narep = string_scalar("",false),
                               rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());

std::unique_ptr<column> prepend(column_view const& string_column,
                                string_scalar const& value,
                                string_scalar const& narep = string_scalar("",false),
                                rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());

in_s = ['aa', null, '', 'aa']

append(in_s, "post", '_')
post_s = ['aapost', '_post', 'post', 'aapost']

prepend(in_s, "pre", '_')
pre_s = ['preaa', 'pre_', 'pre', 'preaa']
```

Creating a column of a single string repeated for every row in the column to append to is an alternative solution, but impractical and slow.

### PR DESCRIPTION
**Pull request by sriramch**
- this Closes #3726
  - this emulates `concatenate_ws` spark functionality
  - provides option for a global separator and global column null replacements
  - skips null values in a row to perform concatenation

### CODE REVIEW

File: cpp/src/strings/combine.cu [hunk 1]
```diff
@@ -245,6 +245,199 @@ std::unique_ptr<column> join_strings(
...
+  CUDF_EXPECTS(num_columns > 0, "At least one column must be specified");
+  // Check if all columns are of type string
+  CUDF_EXPECTS(std::all_of(strings_columns.begin(),
+                           strings_columns.end(),
+                           [](auto c) { return c.type().id() == STRING; }),
+               "All columns must be of type string");
+
+  auto strings_count = strings_columns.num_rows();
+  CUDF_EXPECTS(strings_count == separators.size(),
+               "Separators column should be of the same size of string columns");
```
Comment (davidwendt) [inline cpp/src/strings/combine.cu:266]:
```suggestion
               "Separators column should be the same size as the strings column");
```

Comment (sriramch) [reply cpp/src/strings/combine.cu]: done.

File: cpp/include/cudf/strings/combine.hpp [hunk 2]
```diff
@@ -98,6 +98,64 @@ std::unique_ptr<column> join_strings(
   string_scalar const& narep          = string_scalar("", false),
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Concatenates a list of strings columns using separators for each row
+ * and returns the result as a string column.
```
Comment (davidwendt) [inline cpp/include/cudf/strings/combine.hpp:103]:
```suggestion
 * and returns the result as a strings column.
```

File: cpp/include/cudf/strings/combine.hpp [hunk 3]
```diff
@@ -98,6 +98,64 @@ std::unique_ptr<column> join_strings(
...
+ * out2 is ['aa::-::bb', '-%%cc%%', '^^dd^^-', 'ee!-!-', '-*-*-', null]
+ * @endcode
+ *
+ * @throw cudf::logic_error if no input columns are specified - table view is empty
+ * @throw cudf::logic_error if input columns are not all strings columns.
+ * @throw cudf::logic_error if the number of rows from @p separators and @p strings_columns
+ *                          do not match
+ *
+ * @param strings_columns List of string columns to concatenate.
+ * @param separators String column that provides the separator for a given row
```
Comment (davidwendt) [inline cpp/include/cudf/strings/combine.hpp:142]:
```suggestion
 * @param separators Strings column that provides the separator for a given row
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -29,6 +29,7 @@
 - PR #5149 Add Java bindings for PMOD
 - PR #5153 Add Java bindings for extract
 - PR #5196 Add Java bindings for NULL_EQUALS, NULL_MAX and NULL_MIN
+- PR #5204 Concatenate string columns with row separator column
```
Comment (davidwendt) [inline CHANGELOG.md:32]:
```suggestion
- PR #5204 Concatenate strings columns with row separator column
```

**Comment by sriramch** [comment]
> This doesn't seem to directly address the requirement of the associated issue. I can approve but prefer not to merge this until @rwlee and @revans2 confirm that the alternate approach you mentioned is acceptable.

[sc] i have checked with @revans2 offline and he did concur that the alternate api's requested in the ticket will result in as much (or more) intermediaries compared to existing append api. further, this api is required to support the `concatenate_ws` semantics with a separator per row.

**Comment by revans2** [comment]
This looks fine to me @rwlee does the API look like what you expected?

**Comment by devavret** [comment]
Just out of curiosity, what can this new API achieve that wasn’t possible with the old concatenate. Eg could you achieve the same result by creating a `table_view` from `{col1, separator_col, col2, separator_col, col3}` and an empty string scalar as separator. Is the difference only in the ability to specify different null replacements for value columns and separator columns? 

If so, could this be better achieved by making the existing API more generic and take a null replacement per column of the values table? I’m not prescribing that this is how it should be done, just trying to understand the API requirements.

**Comment by sriramch** [comment]
>> creating a table_view from {col1, separator_col, col2, separator_col, col3} and an empty string scalar as separator. 

- existing api nulls a row even if one column has a null value for the row. this skips nulls
- existing api requires separator rep. to be valid. this doesn't enforce it, and if a separator for a given row is null, the output is null for that row
- this api explicitly defines a separator column vs. having to sneak it in between string columns. the caller has to *make* it to work, which is non-intuitive

>> Is the difference only in the ability to specify different null replacements for value columns and separator columns?

- this is one of them

>> making the existing API more generic and take a null replacement per column of the values table?

- the existing api can be made to work, but i feel it may result in cluttering the existing api. for instance, the existing use-case requires a valid separator scalar. this cannot be enforced, if this has to be retrofitted to the new requirement, as we cannot interpret the semantics of the different columns contained within the table.

**Comment by devavret** [comment]
> * the existing api can be made to work, but i feel it may result in cluttering the existing api

Fair enough

**Comment by rwlee** [comment]
> This looks fine to me @rwlee does the API look like what you expected?

Yup, this addresses the core functionality we wanted. Thanks @sriramch

**Comment by harrism** [comment]
Retargeting to 0.15 since we enter code freeze tonight.

**Comment by harrism** [comment]
Make sure you communicate with a maintainer when you open PRs. This one was not on my radar and therefore had not been added to the 0.14 project board.

**Comment by sriramch** [comment]
@devavret is this good to go?

i have now re-targeted this against 0.15.

**Comment by sriramch** [comment]
@davidwendt @devavret - can one of you please merge this?

**Pull request merged**